include/range/v3/view/for_each.hpp

```diff
@@ -32,61 +32,54 @@ namespace ranges
 {
     inline namespace v3
     {
-        /// \addtogroup group-views
-        /// @{
-        template<typename Rng, typename F>
-        struct for_each_view
-          : join_view<transform_view<Rng, F>>
-        {
-            for_each_view() = default;
-            for_each_view(Rng rng, F f)
-              : for_each_view::join_view{{std::move(rng), std::move(f)}}
-            {}
-        };
-
         namespace view
         {
             struct for_each_fn
             {
             private:
                 friend view_access;
-                template<typename F>
-                static auto bind(for_each_fn for_each, F f)
+                template<typename Fun>
+                static auto bind(for_each_fn for_each, Fun fun)
                 RANGES_DECLTYPE_AUTO_RETURN
                 (
-                    make_pipeable(std::bind(for_each, std::placeholders::_1, protect(std::move(f))))
+                    make_pipeable(std::bind(for_each, std::placeholders::_1, protect(std::move(fun))))
                 )
             public:
-                template<typename Rng, typename F>
-                using Concept = meta::and_<
-                    Range<Rng>,
-                    IndirectInvocable<F, iterator_t<Rng>>,
-                    Range<concepts::Invocable::result_t<F&, range_common_reference_t<Rng>>>>;
-
-                template<typename Rng, typename F,
-                    CONCEPT_REQUIRES_(Concept<Rng, F>())>
-                for_each_view<all_t<Rng>, F> operator()(Rng && rng, F f) const
-                {
-                    return {all(std::forward<Rng>(rng)), std::move(f)};
-                }
+                template<typename Rng, typename Fun,
+                    CONCEPT_REQUIRES_(
+                      transform_fn::Concept<Rng, Fun>()
+                      && InputRange<concepts::Invocable::result_t<
+                          Fun&, range_reference_t<Rng>>>())>
```

Comment:
If `Rng` is not a range, then `range_reference_t<Rng>` is ill-formed and this overload will not be selected. I suggest this:

```c++
template<typename Rng, typename F>
using Concept = meta::and_<
    transform_fn::Concept<Rng, Fun>,
    meta::lazy::invoke<
        meta::compose<
            meta::quote<InputRange>,
            meta::bind_front<meta::quote<concepts::Invocable::result_t>, Fun&>,
            meta::quote<range_reference_t>>,
        Rng>>;
```

Then use `Concept<Rng>()` to constrain the "good" overload and `!Concept<Rng>()` to constrain the "bad" one.